modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  std::vector<Real> angle(n_nodes, 0);
+  std::vector<unsigned int> node_id(_cut_elem_nnode);
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+    angle[i] = 0;
+
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+
+    // To match the type definition and data structure in Xfem::angle_rad_3d
+    // C array is used here instead of std::array or std::vector
+    Real vertices[_cut_elem_nnode][LIBMESH_DIM];
+
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+    {
+      Node * this_node = cut_elem->get_node(i);
+      Point & this_point = *this_node;
+      for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+        vertices[i][j] = this_point(j);
+      node_id[i] = this_node->id();
+    }
+
+    angle[node_id[0]] += Xfem::angle_rad_3d(&vertices[2][0], &vertices[0][0], &vertices[1][0]);
+    angle[node_id[1]] += Xfem::angle_rad_3d(&vertices[0][0], &vertices[1][0], &vertices[2][0]);
+    angle[node_id[2]] += Xfem::angle_rad_3d(&vertices[1][0], &vertices[2][0], &vertices[0][0]);
+  }
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+  {
+    if (std::abs(angle[i] - libMesh::pi * 2) > libMesh::pi * 2 * 0.001)
+    {
+      std::shared_ptr<CutPoint> ptr(new CutPoint);
+      _boundary_map[i] = ptr;
+      Node * this_node = _cut_mesh->node_ptr(i);
+      mooseAssert(this_node != nullptr, "Node is NULL");
+      Point & this_point = *this_node;
+
+      mooseAssert(_boundary_map.find(i) != _boundary_map.end(),
+                  "_boundary_map does not have this key") for (unsigned int j = 0; j < LIBMESH_DIM;
+                                                               ++j) _boundary_map.find(i)
+          ->second->coord[j] = this_point(j);
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::findBoundaryEdges()
+// find boundary edges of the cutter mesh
+{
+  _boundary_edges.clear();
+
+  std::vector<unsigned int> corner_elem_id;
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
+  int counter = -1;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    counter++;
+    Elem * cut_elem = *elem_it;
+    std::vector<unsigned int> node_id(_cut_elem_nnode);
+    std::vector<bool> is_node_on_boundary(_cut_elem_nnode);
+
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+    {
+      node_id[i] = cut_elem->get_node(i)->id();
+      is_node_on_boundary[i] = (_boundary_map.find(node_id[i]) != _boundary_map.end());
+    }
+
+    if (is_node_on_boundary[0] && is_node_on_boundary[1] && is_node_on_boundary[2])
+    {
+      // store corner elements
+      corner_elem_id.push_back(counter);
```

Comment:
You aren't convincing me that you need ints instead of boundary_id type. What if we configure libMesh with a different boundary_id? All this code just fails to compile?

Reply:
Is it fine to use dof_id_type instead of BoundaryID like what's shown below?  If no, is there a clean way to convert between BoundaryID and dof_id_type? static_cast?

std::vector<dof_id_type> _boundary;
std::vector<std::vector<dof_id_type>> _front;
std::map<dof_id_type, std::shared_ptr<CutPoint>> _boundary_map;
std::vector<std::vector<dof_id_type>> _active_boundary;
std::vector<dof_id_type> _inactive_boundary;